src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs

```diff
@@ -15,127 +14,36 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
     /// <typeparam name="TStream"></typeparam>
     internal class DuplexPipeStreamAdapter<TStream> : DuplexPipeStream, IDuplexPipe where TStream : Stream
     {
-        private readonly Pipe _output;
-        private Task _outputTask;
-        private bool _disposed;
-        private readonly object _disposeLock = new object();
-
         public DuplexPipeStreamAdapter(IDuplexPipe duplexPipe, Func<Stream, TStream> createStream) :
             this(duplexPipe, new StreamPipeReaderOptions(leaveOpen: true), new StreamPipeWriterOptions(leaveOpen: true), createStream)
         {
         }
 
-        public DuplexPipeStreamAdapter(IDuplexPipe duplexPipe, StreamPipeReaderOptions readerOptions, StreamPipeWriterOptions writerOptions, Func<Stream, TStream> createStream) :
-            base(duplexPipe.Input, duplexPipe.Output)
+        public DuplexPipeStreamAdapter(IDuplexPipe duplexPipe, StreamPipeReaderOptions readerOptions, StreamPipeWriterOptions writerOptions, Func<Stream, TStream> createStream) : base(duplexPipe.Input, duplexPipe.Output)
         {
             Stream = createStream(this);
-
-            var outputOptions = new PipeOptions(pool: writerOptions.Pool,
-                                                readerScheduler: PipeScheduler.Inline,
-                                                writerScheduler: PipeScheduler.Inline,
-                                                pauseWriterThreshold: 1,
-                                                resumeWriterThreshold: 1,
-                                                minimumSegmentSize: writerOptions.MinimumBufferSize,
-                                                useSynchronizationContext: false);
-
             Input = PipeReader.Create(Stream, readerOptions);
-
-            // We're using a pipe here because the HTTP/2 stack in Kestrel currently makes assumptions
-            // about when it is ok to write to the PipeWriter. This should be reverted back to PipeWriter.Create once
-            // those patterns are fixed.
-            _output = new Pipe(outputOptions);
+            Output = PipeWriter.Create(Stream, writerOptions);
         }
 
-        public ILogger Log { get; set; }
-
         public TStream Stream { get; }
 
         public PipeReader Input { get; }
 
-        public PipeWriter Output
-        {
-            get
-            {
-                if (_outputTask == null)
-                {
-                    _outputTask = WriteOutputAsync();
-                }
-
-                return _output.Writer;
-            }
-        }
+        public PipeWriter Output { get; }
 
-        public override ValueTask DisposeAsync()
+        protected override void Dispose(bool disposing)
```

Comment:
Remove dispose. DisposeAsync should be enough (in my original debugging, base.DisposeAsync will just call Dispose).

Reply:
Wouldn't that make DuplexStreamAdapter.Close()/Dispose() no-op while DuplexStreamAdapter.DisposeAsync() would actually cleanup? That seems bad to me.

https://github.com/dotnet/corefx/blob/ac154845e79c6154c2b502aa3cdd4410e1b5f2b0/src/Common/src/CoreLib/System/IO/Stream.cs#L210-L234

Reply:
If anything, I would remove DisposeAsync() since the default implementation of Stream.DisposeAsync() calls Stream.Dispose(). Do you want me to do that instead?

Reply:
We’re going to make it CompleteAsync eventually though so i rather keep it